### PR TITLE
Allow object editing

### PR DESCRIPTION
**Pull request by nicholette**
Allow objects to be edited after creation. 

~~Commits from #2571  have been added into this PR to show the functionality of this feature. (Will deconflict after #2571 is merged.)~~

To be merged after #2618.

Requires a new scribing answer to test (or there will be errors of `colorPickerColor` not being a `string`)

### CODE REVIEW

File: client/app/bundles/course/assessment/submission/components/ScribingView/ScribingToolbar.jsx [hunk 1]
```diff
@@ -266,13 +266,27 @@ class ScribingToolbar extends Component {
     this.props.setSelectedShape(this.props.answerId, shape);
   }
 
+  setToSelect = () => {
```
Comment (kxmbrian) [inline client/app/bundles/course/assessment/submission/components/ScribingView/ScribingToolbar.jsx]: `setToSelect` is a little confusing. Possible to change to `setToSelectTool`?

**Comment by kxmbrian** [comment]
For the Popovers, could you factor out the component structure such that we can pass in a different hash of styles and callbacks in order to render the active vs normal version?

Also, to make it easier to review, can try to break this PR into more commits? Perhaps something along the lines the following:
- Display border style for `ToolDropdown`
- Allow `ToolDropdown`s to be disabled
- Allow active object to be un/selected
- Re-render canvas and save scribbles when user changes style via popover

File: client/app/bundles/course/assessment/submission/components/ScribingView/ScribingToolbar.jsx [hunk 2]
```diff
@@ -266,13 +266,40 @@ class ScribingToolbar extends Component {
...
       background: this.props.scribing.selectedTool === scribingTools.LINE ? blue500 : 'rgba(0, 0, 0, 0.4)',
     };
     const toolBarStyle = !scribing.isCanvasLoaded ? styles.disabledToolbar : styles.toolBar;
+    let shapeIcon = this.props.scribing.selectedShape === scribingShapes.RECT ?
+      'fa fa-square-o' : 'fa fa-circle-o';
+    if (scribing.activeObject && scribing.activeObject.type === 'rect') {
+      shapeIcon = 'fa fa-square-o';
+    } else if (scribing.activeObject && scribing.activeObject.type === 'ellipse') {
+      shapeIcon = 'fa fa-circle-o';
+    }
```
Comment (kxmbrian) [inline client/app/bundles/course/assessment/submission/components/ScribingView/ScribingToolbar.jsx]: Is it cleaner to do something similar to:

```
const isNewRect = ...
const isEditRect = ..
const shapeIcon = (isNewRect || isEditRect) ? 'fa fa-square-o' : 'fa fa-circle-o';
```

File: client/app/bundles/course/assessment/submission/components/ScribingView/ToolDropdown.jsx [hunk 3]
```diff
@@ -40,44 +42,89 @@ const style = {
...
 
 export default class ToolDropdown extends Component {
   renderIcon() {
-    const { iconClassname, currentTool, toolType, iconComponent } = this.props;
+    const { disabled, iconClassname, currentTool, toolType, iconComponent } = this.props;
+    let iconStyle;
+    if (disabled) {
+      iconStyle = style.disabled;
+    } else if (currentTool === toolType) {
+      iconStyle = { color: blue500 };
```
Comment (kxmbrian) [inline client/app/bundles/course/assessment/submission/components/ScribingView/ToolDropdown.jsx]: To improve readability and consistency, perhaps can do `iconStyle = style.selected`?

File: client/app/bundles/course/assessment/submission/components/ScribingView/ToolDropdown.jsx [hunk 4]
```diff
@@ -40,44 +42,89 @@ const style = {
...
+          width: '23px',
+          height: '8px',
+          background: '#c0c0c0',
+        }}
+      />
+      :
+      <div
+        style={{
+          width: '23px',
+          height: '8px',
```
Comment (kxmbrian) [inline client/app/bundles/course/assessment/submission/components/ScribingView/ToolDropdown.jsx]: Cleaner to write the following? :

```
const colorBarStyle = ...
return <div style={colorBarStyle} />
```

Can make use of `styles.colorBar` to compute `colorBarStyle` also.

File: client/app/bundles/course/assessment/submission/components/ScribingView/ScribingCanvas.jsx [hunk 5]
```diff
@@ -97,17 +100,40 @@ export default class ScribingCanvas extends React.Component {
       }, nextProps.scribing.canvasZoom);
       this.canvas.trigger('mouse:move', { isForced: true });
 
+      // Discard prior active object/group when using other tools
+      if (nextProps.scribing.isChangeTool) {
+        if (nextProps.scribing.selectedTool !== scribingTools.TYPE
+          && nextProps.scribing.selectedTool !== scribingTools.DRAW
+          && nextProps.scribing.selectedTool !== scribingTools.LINE
+          && nextProps.scribing.selectedTool !== scribingTools.SHAPE) {
```
Comment (kxmbrian) [inline client/app/bundles/course/assessment/submission/components/ScribingView/ScribingCanvas.jsx]: Give this bool value a name, maybe 

```
const isNonDrawingTool = nextProps.scribing.selectedTool !== scribingTools.TYPE && ...
```

File: client/app/bundles/course/assessment/submission/components/ScribingView/ScribingToolbar.jsx [hunk 6]
```diff
@@ -283,152 +310,311 @@ class ScribingToolbar extends Component {
...
+            :
+            <TypePopover
+              open={this.state.popovers[scribingPopoverTypes.TYPE]}
+              anchorEl={this.state.popoverAnchor}
+              onRequestClose={() => (this.onRequestClosePopover(scribingPopoverTypes.TYPE))}
+              fontFamilyValue={this.props.scribing.fontFamily}
+              onChangeFontFamily={this.onChangeFontFamily}
+              fontSizeValue={this.props.scribing.fontSize}
+              onChangeFontSize={this.onChangeFontSize}
+              onClickColorPicker={event => (this.onClickColorPicker(event, scribingToolColor.TYPE))}
```
Comment (kxmbrian) [inline client/app/bundles/course/assessment/submission/components/ScribingView/ScribingToolbar.jsx]: Would it be better to factor out common props for each of the Popovers? :
```
const textToolBaseProps = {
    open: this.state.popovers[scribingPopoverTypes.TYPE],
    anchorEl: this.state.popoverAnchor,
    fontFamilyValue: this.props.scribing.fontFamily,
    ... // common items
}

// Merge in other props depending on whether active object is text

<TypePopover {... /* destructure hash */}/>
```

**Comment by kxmbrian** [comment]
The render function for the toolbar is quite long. Is it worth it to factor out each Popover + Tooldrop down into a subcomponent, or at least a separate render function? 

That would have the added benefit of being able to locate the computation of things like `lineToolStyle`, `toolBarStyle`, `shapeIcon` closer to the places where they are actually used.


Perhaps in a later PR, it would be good to refactor the folder structure to reflect the component hierarchy: 

- ScribingView
	- index.jsx
	- ScribingView.scss
	- ScribingToolbar
		- index.jsx
		- SavingIndicator.jsx
		- ToolDropdown.jsx
		- LayersComponent.jsx
		- popovers
			- ...
		- fields
			- ...
	- ScribingCanvas
		- index.jsx
		- ...

It would be helpful then to use absolute paths for imports instead:

```
import { scribingTranslations as translations } from 'course/assessment/submission/translations';
```

File: client/app/bundles/course/assessment/submission/components/ScribingView/ScribingCanvas.jsx (hunk 5 above)
Comment (nicholette) [reply client/app/bundles/course/assessment/submission/components/ScribingView/ScribingCanvas.jsx]: Ok!

**Comment by nicholette** [comment]
@kxmbrian Is it alright if I break down the render method and do the major refactoring in another PR? Just in case it breaks the existing code.

File: client/app/bundles/course/assessment/submission/components/ScribingView/ToolDropdown.jsx [hunk 7]
```diff
@@ -40,44 +42,79 @@ const style = {
...
+
+    const colorBarStyle = disabled ?
+    {
+      width: '23px',
+      height: '8px',
+      background: '#c0c0c0',
+    } :
+    {
+      width: '23px',
+      height: '8px',
```
Comment (kxmbrian) [inline client/app/bundles/course/assessment/submission/components/ScribingView/ToolDropdown.jsx]: If you prefer not to use `style.colorBar`, can remove it? It doesn't seem to be used anywhere

**Comment by kxmbrian** [comment]
Sure, next PR is fine. Thanks for making the changes. It made this PR easier to check.

Another thing is that if you draw say a circle, you can't finalise it by clicking the select tool (changes will be lost). Is it a good idea to have that behaviour? If yes, can make the change in a separate PR.

After the last housekeeping item, I think we can merge this. Thanks for you patience!

**Comment by nicholette** [comment]
@kxmbrian Eh nope, that behaviour was not intended, didn't catch that. Thanks!

**Pull request merged**